streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/ForeignTableJoinProcessorSupplier.java

```diff
@@ -111,34 +110,26 @@ public void process(final Record<KO, Change<VO>> record) {
                 return;
             }
 
-            final Bytes prefixBytes = keySchema.prefixBytes(record.key());
+            final Bytes foreignKeyBytes = keySchema.prefixBytes(record.key());
 
             //Perform the prefixScan and propagate the results
             try (final KeyValueIterator<Bytes, ValueAndTimestamp<SubscriptionWrapper<K>>> prefixScanResults =
-                     subscriptionStore.range(prefixBytes, Bytes.increment(prefixBytes))) {
+                     subscriptionStore.prefixScan(foreignKeyBytes, new BytesSerializer())) {
```

Comment:
note that this isn't totally backwards compatible since some state stores don't implement `prefixScan` (while all stores need to implement `range`) -- the default throws an `UnsupportedOperationException`.

I'm also not certain it's totally correct. What if I have a foreign key that is `aaa` and then a foreign key that is `aaab`. The prefix scan would return records matching `aaab`.

We should at minimum add tests to make sure this is covered.

Reply:
I think we can make this backwards compatible by implementing a poor-man's prefix scan as the default implementation for `prefixScan` using the same strategy the original code had (by checking equality on every retrieved prefix to make it exclusive)

Reply:
I think we can implement an alternative for `Bytes.increment(prefixBytes)` such that `range` always return the correct results.

Reply:
@aliehsaeedii which kind of alternative?

Reply:
I have another idea: as ranges should be ordered, we could just check the last key found instead of all

Reply:
@Chuckame 
Something like this may work?!
```
private Bytes myIncrement(Bytes input) {
    byte[] inputArr = input.get();
    byte[] ret = new byte[inputArr.length + 1];
    int i;
    for (i = 0; i < inputArr.length; i++) {
        ret[i] = inputArr[i];
    }
    ret[i] = (byte) 0xFF;
    return wrap(ret);
}
```

Reply:
This makes sense to me! Great idea @aliehsaeedii, we need to think it through a bit deeper but I think this works.

Reply:
The main issue is that replacing the current default `prefixScan` is complicated as we are not talking about `Bytes` but generic types:
```java
    default <PS extends Serializer<P>, P> KeyValueIterator<K, V> prefixScan(P prefix, PS prefixKeySerializer) {
        throw new UnsupportedOperationException();
    }
```

We could still check if the `prefix` is of type `Bytes` and if `prefixKeySerializer` is about `BytesSerializer`... I don't have any other solution to replace this `UnsupportedOperationException`...

Reply:
thanks @Chuckame. Could the solution suggested in these comments ([1](https://github.com/apache/kafka/pull/16218#discussion_r1784582572) and [2](https://github.com/apache/kafka/pull/16218#discussion_r1789789228)) work?

Reply:
Personally I prefer to use the native prefixScan instead of a range "hack" to profit the RocksDB native implementation, as it's using the index and filters differently, and may load into memory different part of the index.

I'm not sure if a range using your increment would work, as the range is inclusive, so we are not able to predict what's the maximum value before the increment.

Taking the following keys `AAA -> ZZZ` (where we only have A-Z letters), if we prefixScan for `A`, the inclusive-end should be the last value before `BAA` which is `AZZ`. Your increment suppose to compute a range from `A` to `AZ`, but it will only take values until `AZ`, while `AZZ` is greater than `AZ`. Comparatively: `A < AB < AXZ < AZ < AZA < AZZ`


I think the easiest solution is to just exclude the entry when the key is equals to the incremented bytes, but only for the last key instead of all the entries, as a range is ordered:
```diff
//Perform the prefixScan and propagate the results
            try (final KeyValueIterator<Bytes, ValueAndTimestamp<SubscriptionWrapper<K>>> prefixScanResults =
                         subscriptionStore.range(foreignKeyBytes, Bytes.increment(foreignKeyBytes))) {

                while (prefixScanResults.hasNext()) {
                    final KeyValue<Bytes, ValueAndTimestamp<SubscriptionWrapper<K>>> next = prefixScanResults.next();
                    // have to check the prefix on the last key because the range end is inclusive :(
-                    if (prefixEquals(next.key.get(), foreignKeyBytes.get())) {
+                    if (prefixScanResults.hasNext() || prefixEquals(next.key.get(), foreignKeyBytes.get())) {
                        final CombinedKey<KO, K> combinedKey = keySchema.fromBytes(next.key);
                        context().forward(
                                record.withKey(combinedKey.primaryKey())
                                        .withValue(new SubscriptionResponseWrapper<>(
                                                next.value.value().hash(),
                                                record.value().newValue,
                                                next.value.value().primaryPartition()))
                        );
                    }
                }
            }
```

Reply:
Btw, I'm still unsure if we accept this kind of unsafe casts in the default poor's man implementation, where we assume that the prefix must be `Bytes` and the key should be also `Bytes` (as we use `range(prefix, Bytes.increment(prefix))`:

```java
    default <PS extends Serializer<P>, P> KeyValueIterator<K, V> prefixScan(P prefix, PS prefixKeySerializer) {
        if (!(prefix instanceof Bytes)) {
            throw new UnsupportedOperationException();
        }
        final KeyValueIterator<K, V> inclusiveRangeIterator = range((K) prefix, (K) Bytes.increment((Bytes) prefix));
        return (KeyValueIterator<K, V>) new PrefixScanKeyValueIterator<>((KeyValueIterator<Bytes, V>) inclusiveRangeIterator, (Bytes) prefix);
    }
```

Isn't it a candidate of breaking change for 4.0 client ?

Reply:
@AyoubOm 
By `if (prefixScanResults.hasNext() || prefixEquals(next.key.get(), foreignKeyBytes.get()))` I assume you mean                `if (next.hasNext() || prefixEquals(next.key.get(), foreignKeyBytes.get()))`. Anyways, I think foreign keys are not unique, therefore, a key which is out of range (`b` for example while we are interested only in `a`) may have further elements in the iterator.

I tried to run the unit tests and I noticed in [this test](https://github.com/apache/kafka/blob/cebec91470c440a1735e50210ae3fc0f62ee40a7/streams/src/test/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/ForeignTableJoinProcessorSupplierTests.java#L96-#L113) for example, the range is from `fk1` to `Bytes.increment(prefixBytes)` which is `fk2`. The output of `range` contains only the records with `fk1` keys, even if other records with `fk2` or `fk11` has been added to the store!

Reply:
cc: @Chuckame

Reply:
@aliehsaeedii I don't want to be pushy, but you did not answer my remark/message. Also, I'm not sure about what do you want to achieve regarding your answer to @AyoubOm.

In any case, there is an issue, as we miss the concrete type in the store interface.

Reply:
I am sorry @Chuckame. I wanted to ping you and mistakenly `AyoubOm` was pinged. I am only communicating with you here since you are the PR owner.